### ISSUE TITLE
run npmbox on a local, unpublished package

### ISSUE THREAD

**Issue by ldegen** [arei/npmbox#31]
Hi,

It seems that `npmbox` requires the package to be published before it can be bundled.

Do you think this could be changed? I.e. would it be possibe to make `npmbox` work on the current directory, similar to the way `npm` does?

**Comment by arei** [comment]
I understand what you are shooting for but it's a little bit outside of the scope of npmbox. It would almost certainly be an entirely different process to acheive than npmbox uses. Almost a separate utility would be better.

All that said, this definately would not be a priority for me or npmbox.  I welcome pull requests though.

**Comment by ldegen** [comment]
Hm... maybe I could try and pull off some sort of npm registry mockery...
I'll have to do some more research. 
Thanks for the reply, though.

**Comment by danfuzz** [comment]
I did some investigation into this today, and I think I've figured out that this feature won't actually be too terribly hard to add. In particular, `npm cache add` will already recognize `file:...` to indicate a local file. Names of this form become `packageType === "local"`, which can be checked for and acted upon similarly to types `git` and `hosted`, e.g. at <https://github.com/arei/npmbox/blob/master/npmboxxer.js#L248>.

Two gotchas I noticed:

* The report of `Package has no name` at <https://github.com/arei/npmbox/blob/master/npmboxxer.js#L256> is called even when a package does have a name. I think the line might be just missing an `else`.

* The function `rack` at <https://github.com/arei/npmbox/blob/master/npmboxxer.js#L223> wants to construct a file name that includes the source name, which in this case can contain slashes and therefore confuse things (e.g. `.npmbox_cache/file:../../my-local-package`). I couldn't figure out how these files are used, but maybe it would suffice to encode away the slashes in the name (e.g. percent-encoding a la URLs).

I didn't get to the point of actually having a PR to submit (sorry). Nonetheless, I hope this investigation helps y'all.

**Comment by danfuzz** [comment]
Hi! Me again. I spent the day trying to add this feature for real, and I believe I'm mostly done (still need to do a bunch of testing before I'll be happy with it). You can find the branch at <https://github.com/danfuzz/npmbox/tree/local-package>.

I added comments around anything that I thought might not be clear, and I tried to provide a reasonable commit history. If you have the time to look at it I'd be most appreciative.

BTW, whether or not you want this as a PR in its entirety, you might still be interested in the commit <https://github.com/danfuzz/npmbox/commit/1411414217bfdb2e732851deb46e0cd42d4e0511> (or something like it) which works around a problem with the `tar.gz` package.

**Comment by arei** [comment]
Mostly like but some thoughts...

1. Instead of the stripping outtp the characters that don't work in the file names, what about just URL encoding them? That keeps them if we need them later... kind of future proofing if you will.  Just a thought.

2. In addition to doing the encoding, do we need to do a basename as well to just get the package name we want without all the path stuff? Just wondering.

3. In npmboxxer.js the line on 380 that you comment out is required for backwards compatibility with older npmbox files. In addition to commenting out I think we need to check after we are done reading the contents of the tar to see if there were reference pointers in the tar. If not, we need to fall back to the original name as what we are installing.  Did that make sense?

Otherwise great work. I look forward to your pull request.

**Comment by danfuzz** [comment]
Hi! Thanks for the quick response. I'll keep working on this today. Responses to your items in-order:

1. I actually did this in my original prototype and ended up ditching it because (believe it or not) I expected you would think it was overkill. I'd be happy to rejigger it, but see below for a longer discussion.

2. The regex that I wrote gets both the scheme (`git:` or etc.) and the final path component. I could use `basename` for the final component, but I don't know that that's really particularly clearer than what I wrote. I could maybe use a URL parser library? I defer to you. But also maybe it's moot (again see below).

3. Yeah totally. Thanks for the history. I'll fix it.

Aforementioned longer discussion:

It took me a while to understand how the pieces fit together, and since my comment yesterday I think I've achieved a more nuanced understanding of what's going on.

The way things stand in my branch, that URL-mangling code doesn't even get activated, because of how I changed `rack()` to take the package name, and I believe that's the right arrangement. That is, I think the right thing is to remove the URL-mangling code and leave the rest at least mostly as-is. Here's why:

(As you know but I didn't quite get before…) The point of the flag files is to provide the "roots" for what to ask `npm` to install. Once in an `.npmbox` archive, anything that isn't a regular package (hosted, git, local) is still just represented in the directory structure as `name/version`, so the flag file just needs to point to the name such that `npm` will find it. And the thing `npm` will succeed with at that point is just the package name, not the URL (or local file path, etc.).

One tweak that strikes me as worthwhile is making the flag file for hosted (etc.) packages point explicitly at the version of the package that got included. E.g. the flag file name would be something like `pkg-name@1.2.3.npmbox` instead of just `pkg-name`. That way, in the (admittedly edgiest of edge) case where you have more than one version of a package with the same name, there won't be ambiguity about which one is the "root."

Thanks again, and let me know what you think. I'll do my best to encode your suggestions in the mean time.

**Comment by danfuzz** [comment]
BTW, I don't know what your preferences are for maximum line width (and relatedly what your tab width is). Happy to adjust the code to suit.

**Comment by danfuzz** [comment]
Afternoon update: I believe I have the boxing side of things all working now, a bit simplified from the state of affairs yesterday evening. Unfortunately, on the unboxing side, there's _something_ in a cached local package which makes `npm` think that it's invalid and so tries to query it and fetch it from npm central. I don't yet know what's going on.

All current work has been pushed to the branch. As before, please feel free to check it out, comment, etc.

**Comment by arei** [comment]
Tabs over space.  4 spaces to a tab.  No worries if you deviant, will just readjust later when I need too. 

On the afternoon update... there may be a strange behavior with actual file names cuz and packages in the tar named xyz.  See some of the other open issues for context. Don't know if it relates, just a strange intuition I have.

**Comment by danfuzz** [comment]
Thanks for the info. Still not sure what's going on, but I _do_ notice that there's a bunch of cached registry info under `<cache-dir>/registry.npmjs.org/<package-name>/.cache.json` _and_ I noticed that a local package shows up as being `_from` version `latest` and nothing more specific (this is a synthetic npm-generated field in cached `package.json` files). These two facts may be interplaying in npm's decision to try querying the registry.

BTW, I've been testing by boxing `npmbox` itself, and I only just thought to try boxing / unboxing it as published, using the current (not my modified) version of the tool. It currently doesn't work for me, which is kinda surprising. I can repro by "manually" running the same `npm install` command that `npmbox` does (after untarring the box file, etc.). Looks like npm decides it needs to check the registry for `npmbox` and for whatever reason the cached info isn't good enough, not entirely unlike what I'm seeing with a local package. I've been pretty careful about not polluting my test spaces, but I'm not 100% sure there isn't some leakage. In any case, salient transcript:

```
$ npm cache clear
$ ../tool/node_modules/.bin/npmbox npmbox # The code in `tool` is a vanilla `npm install npmbox`.
Boxing npmbox...
  Querying npmbox
  Querying is@3.1.0
  ...
  Packing /Users/danfuzz/testing/zzz3/npmbox.npmbox...
$ [Turned off networking here.]
$ ../tool/node_modules/.bin/npmunbox npmbox.npmbox 
Unboxing npmbox.npmbox...
  Unpacking npmbox.npmbox...
  Installing npmbox...
An error occurred while installing npmbox.
npmbox was not installed.
  Done.
$ tar xzf npmbox.npmbox 
$ npm --cache=`pwd`/.npmbox_cache --cache-min=99999 --fetch-retries=0 --fetch-retry-factor=0 --fetch-retry-mintimeout=1 --fetch-retry-maxtimeout=2 --ignore-scripts install npmbox
...
npm ERR! network getaddrinfo ENOTFOUND registry.npmjs.org registry.npmjs.org:443
...
$ cat npm-debug.log
...
9 silly mapToRegistry name npmbox
10 silly mapToRegistry using default registry
11 silly mapToRegistry registry https://registry.npmjs.org/
12 silly mapToRegistry data Result {
12 silly mapToRegistry   raw: 'npmbox',
12 silly mapToRegistry   scope: null,
12 silly mapToRegistry   escapedName: 'npmbox',
12 silly mapToRegistry   name: 'npmbox',
12 silly mapToRegistry   rawSpec: '',
12 silly mapToRegistry   spec: 'latest',
12 silly mapToRegistry   type: 'tag' }
13 silly mapToRegistry uri https://registry.npmjs.org/npmbox
14 verbose request uri https://registry.npmjs.org/npmbox
15 verbose request no auth needed
16 info attempt registry request try #1 at 1:05:38 PM
17 verbose request id bde6888fc98a33e4
18 http request GET https://registry.npmjs.org/npmbox
...
```

**Comment by arei** [comment]
Again, I think this is related to there being a "npmbox" in the tar and in the filesystem. For whatever reason that cause npm to not see the cache folder.

**Comment by danfuzz** [comment]
Oh sorry, I didn't get what you meant before.

**Comment by danfuzz** [comment]
I think I've convinced myself that at least some of the network requests are happening because `npmbox` has a shrinkwrap file. (What I did: Modify the shrinkwrap file in the cached package to have a tell-tale URL and watch `npm` attempt to fetch same.)

**Comment by danfuzz** [comment]
"I'm in a mazy of twisty passages…"

I gave it a good run, but I've been unable to make further headway today. I'm about to be off for the (US) holiday, which means I probably won't get a chance to make any more progress this week, and next week is short for me too. Sorry I don't have a PR for you. Per previous, you might want to pull in <https://github.com/danfuzz/npmbox/commit/1411414217bfdb2e732851deb46e0cd42d4e0511> by itself, as it does fix an independent issue.

Cheers.

**Comment by arei** [comment]
No worries on getting it done.  Go enjoy your turkey.

**Comment by danfuzz** [comment]
Per my work on PR #68, it turns out that — as you guessed — having the full install URL turns out to be necessary to pass to the unboxing invocation of `npm`; otherwise it won't find the cached material in the box.

**Comment by danfuzz** [comment]
If #68 gets merged, then the remaining work to address this bug turns out to be pretty minimal. I made a new branch for it at <https://github.com/danfuzz/npmbox/tree/local-package-ng>.

**Comment by danfuzz** [comment]
…aaaaaand of course I'm wrong about that "minimal" part. Turns out that `npm` won't find cached local packages (which raises the question of why it allows one to cache them at all). I'll keep on truckin'.

**Comment by danfuzz** [comment]
I just updated the branch with one additional commit, which really truly gets local packages working _if they are the direct targets to install_. I'm pretty sure this is the overwhelmingly dominant use case of local packages (and in fact is the one that I need for my own work).

I believe it will take a fair bit more work to make local packages that depend on other local packages, because that will involve adding a pre-install pass over the cache to (a) move cached local package files out of the cache (to convince `npm` that they are okay to install) and (b) rewrite `package.json` files with local package paths to point at those moved files (instead of having their original fs paths which of course don't necessarily exist on the machine where the unboxing takes place).

Anyhoo, at this point I'm pretty happy with the whole work arc and would be excited to see it published as `npmbox` v4.0.2 if you're game. As before, happy to have comments, suggestions, etc. 

Thanks for taking the time to deal with this!

**Comment by danfuzz** [comment]
ICYC I re-re-rebased the local package work, now based off of PR #71. Find it at <https://github.com/danfuzz/npmbox/tree/local-package-take-3>. Same caveat as before.

Still hoping for comments, suggestions, and — natch — a merge. Thanks in advance!

**Comment by ldegen** [comment]
For the sake of completeness and in reference to my original request: In the last couple of days I did a little hack that kind of works __for me__ and __for now__: https://github.com/ldegen/bigpack

The basic idea is: create a new package.json in a temp directory, have it depend on the original one, find transitive closure of all production dependencies and mark them as 'bundledDependencies' in the new package. Then use `npm pack` to create a tar-ball of that.

I would still love a more "complete" solution integrated with npmbox.

**Comment by ldegen** [comment]
regarding my last comment: ignore that. it doesn't work at all. Sorry.

**Comment by danfuzz** [comment]
This should be fixed with v4.1.0, just published.
